rfcs/0154-Migrate-taskcluster-to-postgres.md

```diff
@@ -0,0 +1,165 @@
+bunch of Azure tables to handle tasks, dependencies, artifacts, and so on. This
+project involves refactoring Taskcluster to use a Postgres backend as the
+data-storage backend, instead.
+
+# Motivation
+* Azure doesn't allow joins. Postgres allows more sophisticated querying.
+* Postgres can index by other than the primary key.
+  * A shared base of indexed data can allow services to make smarter decisions and
+    save money on provisioning and data storage.
+* Azure is unreliable with frequent failures and inscrutible rate-limiting. While we are resilient
+to these errors, they do cause inefficiency and sometimes cascading failures.
+* Storing data in postgres will enable features users desire:
+  * Queue introspection (Azure doesn't allow introspection of queues).
+    * What tasks are pending right now?
+    * How many tasks are [actually pending and not cancelled](https://bugzilla.mozilla.org/show_bug.cgi?id=1434851)?
+  * Integration of task-related and worker-related status.
+    * The approach we've taken with Azure doesn't allow tables to read from
+      other tables.
+    * What tasks has this worker performed?
+  * Ad-hoc analytics.
+    * How much compute time do we use per push?
+    * What do we pay for wpt?
+  * More sophisticated dependency resolution [in the case of task failure](https://bugzilla.mozilla.org/show_bug.cgi?id=1443503#c8)
+  * Dynamically change task priorities.
+
+# Details
+
+## Transition
+
+The transition to Postgres will be done in two steps:
+1. Transfer the data from Azure to Postgres without data loss. This will not aim
+   at normalizing the data but rather prioritize a safe and easy transition.  In
+   other words, the data will be backed by Postgres, but still behaving like
+   Azure.
+2. Go fully into normalized SQL tables.
+
+In both steps, Taskcluster will get confidence in the correctness and
+performance of the implementation by load testing it on the development and
+staging environments prior to running this on the production environment.
+
+Upon completion of step 1, deployers of Taskcluster will need to provide a DB
+URL (maybe several) then run a script to "load" data from Azure to Postgres
+during a downtime.
+
+There will be a version of Taskcluster that supports Postgres and not Azure,
+with the previous version supporting Azure and not Postgres. So upgrading to
+that version is something that happens in a downtime.
+
+The point of no return will be 2 weeks. This should give us enough time to see
+if anything goes wrong. At any time before the point of no return, rolling back
+will consist of switching back to a previous release of Taskcluster and
+reverting any kubernetes config changes that were done for this migration. The
+impact of rolling back will be data loss of all data after the planned downtime.
+There will be no data loss if rolling back happens during the downtime.
+
+### Tracing
+
+Taskcluster will use New Relic to a have better visibility of the database,
```

Comment:
Can this be expanded on? We haven't used New Relic with tc before. What sort of changes do we need to make to support it. Why New Relic instead of something else, etc.

Not saying I disagree, just want to know more.

Reply:
To support it i think you'd want to add an env var to conditionally load the [newrelic](https://www.npmjs.com/package/newrelic) package, so we can turn it on in stage and prod but not in local dev. You'll also need to add a couple env vars for config. Beyond that, _hopefully_ it will monkeypatch the relevant libraries like `pg` and just work™ 

New Relic was my recommendation because it's something Mozilla has licenses for and some other teams use it extensively. https://cloud.google.com/trace/ plus https://googleapis.dev/nodejs/trace/latest/ would be a potential alternative.

The goal of using a tracing / apm service during the migration is to get clearer visibility into query performance over time than we can get from application logging or pg views and logs alone.